python/cudf_polars/cudf_polars/callback.py

```diff
@@ -5,43 +5,139 @@
 
 from __future__ import annotations
 
+import contextlib
 import os
 import warnings
-from functools import partial
+from functools import cache, partial
 from typing import TYPE_CHECKING
 
 import nvtx
 
 from polars.exceptions import PerformanceWarning
 
+import rmm
+from rmm._cuda import gpu
+
 from cudf_polars.dsl.translate import translate_ir
 
 if TYPE_CHECKING:
+    from collections.abc import Generator
+
     import polars as pl
+    from polars import GPUEngine
 
     from cudf_polars.dsl.ir import IR
     from cudf_polars.typing import NodeTraverser
 
 __all__: list[str] = ["execute_with_cudf"]
 
 
+@cache
+def default_memory_resource(device: int) -> rmm.mr.DeviceMemoryResource:
+    """
+    Return the default memory resource for cudf-polars.
+
+    Parameters
+    ----------
+    device
+        Disambiguating device id when selecting the device. Must be
+        the active device when this function is called.
+
+    Returns
+    -------
+    rmm.mr.DeviceMemoryResource
+        The default memory resource that cudf-polars uses. Currently
+        an async pool resource.
+    """
+    return rmm.mr.CudaAsyncMemoryResource()
```

Comment:
Should we try to initialize a resource/do a cuda operation on import?

(This operation has to go in __init__.py before the rest of cudf_polars is imported)

Otherwise, the import of cudf_polars will crash if you don't have CUDA12.

The alternative is we can set the ``RAPIDS_NO_INITIALIZE`` env var.
We still get a noisy warning from the numba failed init, though.

```
>>> a = pl.DataFrame({"a": [1,2,3]}).lazy().collect(engine="gpu")
/data/tholi/miniconda3/envs/polars-test/lib/python3.10/site-packages/cudf/utils/_ptxcompiler.py:64: UserWarning: Error getting driver and runtime versions:

stdout:



stderr:

Traceback (most recent call last):
  File "<string>", line 7, in <module>
  File "/data/tholi/miniconda3/envs/polars-test/lib/python3.10/site-packages/numba/cuda/cudadrv/runtime.py", line 111, in get_version
    self.cudaRuntimeGetVersion(ctypes.byref(rtver))
  File "/data/tholi/miniconda3/envs/polars-test/lib/python3.10/site-packages/numba/cuda/cudadrv/runtime.py", line 65, in __getattr__
    self._initialize()
  File "/data/tholi/miniconda3/envs/polars-test/lib/python3.10/site-packages/numba/cuda/cudadrv/runtime.py", line 51, in _initialize
    self.lib = open_cudalib('cudart')
  File "/data/tholi/miniconda3/envs/polars-test/lib/python3.10/site-packages/numba/cuda/cudadrv/libs.py", line 65, in open_cudalib
    return ctypes.CDLL(path)
  File "/data/tholi/miniconda3/envs/polars-test/lib/python3.10/ctypes/__init__.py", line 374, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: libcudart.so: cannot open shared object file: No such file or directory


Not patching Numba
  warnings.warn(msg, UserWarning)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/data/tholi/miniconda3/envs/polars-test/lib/python3.10/site-packages/polars/lazyframe/frame.py", line 1993, in collect
    return wrap_df(ldf.collect(callback))
polars.exceptions.ComputeError: RuntimeError: RMM failure at: /tmp/pip-build-env-svzuti84/normal/lib/python3.10/site-packages/librmm/include/rmm/mr/device/cuda_async_memory_resource.hpp:94: cudaMallocAsync not supported with this CUDA driver/runtime version
```

Reply:
Definitely not on _import_, nothing should initialize a cuda context on import, since that breaks the world in a multi-gpu setting. So thanks, we should set `RAPIDS_NO_INITIALIZE` in the environment before importing anything in cudf polars.

Reply:
I did something that avoids import-time initialisation and hides the warnings